dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
+              static_cast<IUnknown *>(m_pEnum), hr);
+    }
+
+    // update ac list if necessary
+    if (bExpand && m_pACList)
+    {
+        HRESULT hr = m_pACList->Expand(strStemText); // IACList::Expand
+        TRACE("m_pACList->Expand(%p, %S): 0x%08lx\n",
+              static_cast<IUnknown *>(m_pACList),
+              static_cast<LPCWSTR>(strStemText), hr);
+    }
+
+    if (bExpand || m_innerList.GetSize() == 0)
+    {
+        // reload the inner list
+        ReLoadInnerList();
+    }
+
+    return m_innerList.GetSize();
+}
+
+INT CAutoComplete::UpdateOuterList()
+{
+    CStringW strText = GetEditText(); // get the text
+
+    // update the outer list from the inner list
+    m_outerList.RemoveAll();
+    for (INT iItem = 0; iItem < m_innerList.GetSize(); ++iItem)
+    {
+        const CStringW& strTarget = m_innerList[iItem];
+        if (strTarget != strText &&
+            ::StrCmpNIW(strTarget, m_strText, strText.GetLength()) == 0)
+        {
+            m_outerList.Add(strTarget);
+        }
+    }
+
+    // sort the list
+    DoSort(m_outerList);
+    // unique
+    DoUniqueAndTrim(m_outerList);
+
+    // set the item count of the virtual listview
+    INT cItems = m_outerList.GetSize();
+    if (strText.IsEmpty())
+        cItems = 0;
+    m_hwndList.SendMessageW(LVM_SETITEMCOUNT, cItems, 0);
+
+    return cItems; // the number of items
+}
+
+VOID CAutoComplete::UpdateCompletion(BOOL bAppendOK)
+{
+    TRACE("CAutoComplete::UpdateCompletion(%p, %d)\n", this, bAppendOK);
+
+    // update inner list
+    UINT cItems = UpdateInnerList();
+    if (cItems == 0) // no items
+    {
+        HideDropDown();
+        return;
+    }
+
+    if (CanAutoSuggest()) // can we auto-suggest?
+    {
+        m_bInSelectItem = TRUE; // don't respond
+        SelectItem(-1); // select none
+        m_bInSelectItem = FALSE;
+
+        if (UpdateOuterList())
+            RepositionDropDown();
+        else
+            HideDropDown();
+        return;
+    }
+
+    if (CanAutoAppend() && bAppendOK) // can we auto-append?
+    {
+        DoAutoAppend();
+        return;
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// CAutoComplete message handlers
+
+// WM_CREATE
+// This message is sent when the window is about to be created after WM_NCCREATE.
+// The return value is -1 (failure) or zero (success).
+LRESULT CAutoComplete::OnCreate(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CAutoComplete::OnCreate(%p)\n", this);
+
+    // set the pointer of CAutoComplete
+    m_hwndList.m_pDropDown = this;
+    m_hwndScrollBar.m_pDropDown = this;
+    m_hwndSizeBox.m_pDropDown = this;
+
+    // create the children
+    m_hwndList.Create(m_hWnd);
+    if (!m_hwndList)
+        return -1; // failure
+    m_hwndSizeBox.Create(m_hWnd);
+    if (!m_hwndSizeBox)
+        return -1; // failure
+    m_hwndScrollBar.Create(m_hWnd);
+    if (!m_hwndScrollBar)
+        return -1; // failure
```

Comment:
Why is this needed to create all this? Don't they come automatically when using the combobox control? (and then, only retrieving the corresponding existing controls when they show up?)

Reply:
WinHier says these controls should be created.

Reply:
Yes indeed, I've checked and can confirm, it's actually a hidden listview with scrollbar + resizing, that gets shown whenever autocompletion is needed, and just hidden when it's not (and created / destroyed, when an editable window that needs autocompletion is created/destroyed).